service/src/main/java/org/cbioportal/service/impl/GenesetCorrelationServiceImpl.java

```diff
@@ -0,0 +1,264 @@
+        			return -1;
+        		
+        		return 0;
+            }
+        };
+		Collections.sort(result, comparator);		
+	}
+
+
+	/**
+	 * Prepares the gene values, but also filters both them and genesetValues, removing samples where the value is not present in 
+	 * either gene or gene set dimension. 
+	 * 
+	 * @param sampleIds
+	 * @param geneticDataItems
+	 * @param genesetValues
+	 * @return
+	 */
+	private FilteredGeneAndGenesetValues getAndFilterValues(List<String> sampleIds, List<GeneGeneticData> geneticDataItems, double[] genesetValues) {
+		
+		//index geneData values
+		Map<String, Double> sampleValues = new HashMap<String, Double>();
+		for (GeneGeneticData geneticDataItem : geneticDataItems) {
+			double value = Double.NaN;
+			if (NumberUtils.isNumber(geneticDataItem.getValue())) {
+				value = Double.parseDouble(geneticDataItem.getValue());
+			}
+			sampleValues.put(geneticDataItem.getSampleId(), value);
+		}
+		//get values
+		List<Double> geneValueList = new ArrayList<Double>();
+		List<Double> genesetValueList = new ArrayList<Double>();
+		for (int i = 0; i < sampleIds.size(); i++) {
+			String sampleId = sampleIds.get(i);
+			Double value = sampleValues.get(sampleId);
+			if (value == null) {
+				value = Double.NaN; //set to NaN when value is not available for this sample
+			}
+			// if both this value and genesetValues[i] are NaN, then skip this item (since it will not be allowed by the correlation method):
+			if (!Double.isNaN(value) && !Double.isNaN(genesetValues[i])) {
+				geneValueList.add(value);
+				// build up genesetValues is same way:
+				genesetValueList.add(genesetValues[i]);
+			}
+		}
+		// final result:
+		FilteredGeneAndGenesetValues filteredGeneAndGenesetValues = new FilteredGeneAndGenesetValues(); 
+		filteredGeneAndGenesetValues.geneValues = geneValueList.stream().mapToDouble(d -> d).toArray();
+		filteredGeneAndGenesetValues.genesetValues = genesetValueList.stream().mapToDouble(d -> d).toArray();
+		
+		return filteredGeneAndGenesetValues;
+	}
+
+
+	private double[] getGenesetValues(List<String> sampleIds, List<GenesetGeneticData> genesetDataItems) {
+
+		//index genesetData values
+		Map<String, Double> sampleValues = new HashMap<String, Double>();
+		for (GenesetGeneticData genesetDataItem : genesetDataItems) {
+			double value = Double.NaN;
+			if (NumberUtils.isNumber(genesetDataItem.getValue())) {
+				value = Double.parseDouble(genesetDataItem.getValue());
+			}
+			sampleValues.put(genesetDataItem.getSampleId(), value);
+		}
+		//get values
+		double[] result = new double[sampleIds.size()];
+		for (int i = 0; i < sampleIds.size(); i++) {
+			String sampleId = sampleIds.get(i);
+			Double value = sampleValues.get(sampleId);
+			if (value != null) {
+				result[i] = value;
+			} else {
+				result[i] = Double.NaN; //set to NaN when value is not available for this sample
+			}
+		}
+		return result;
+	}
+	
+	class FilteredGeneAndGenesetValues {
```

Comment:
Yeah, you are right, we wouldn't want `model` to be too crowded either. I think a Map would be better here, as this class is not really part of the domain and is just a temporary value holder.